### ISSUE TITLE
WebView source overlay?

### ISSUE THREAD

**Issue by philn** [voc/voctomix#242]
Hi!
I've developed a GStreamer source element leveraging WPEWebKit to provide video frames representing a web page to downstream elements of the pipeline. I talked about it at FOSDEM, see https://fosdem.org/2019/schedule/event/media_gstreamer_wpe/

Would you be interested in supporting such use-case?

**Comment by nueces** [comment]
Yes, I've been waiting for this for a long time! :D

**Comment by MaZderMind** [comment]
That would be a great addition if shipped together with an example web-application that can, for example, render [lower-thirds](https://en.wikipedia.org/wiki/Lower_third) ("Bauchbinden").

I would imagine, one would start a separate web-server-app which presents a "view" and an "control" endpoint. The view-endpoint would be configured into voctomix, the control-endpoint could be opened on a tablet/smartphone/computer to control content, style, placement and visibility/animation of the lower third.

**Comment by nueces** [comment]
There is a scoreboard used for Roller Derby that have a few overlays to use in streaming transmission that i will love to use with this. That is my user case :)

[0] https://github.com/rollerderby/scoreboard

**Comment by fightling** [comment]
Hey @philn.

I'm looking forward to watch your talk later today. But here's something I want to say just now:

@danimo is currently working on an overlay channel which provides alpha channel transparency mixing. We had great success with a first approach yesterday at CCCB Berlin. Source of the overlay could be anything which produces AYUV or ARGB. Rendering the content is still an open issue - especially when we talk about lower-thirds where the input data could be fetched from a web server (like name of speaker etc.).

I think your expertise could help us a lot to do things right here. But let me watch your talk...

**Comment by derpeter** [comment]
> Rendering the content is still an open issue - especially when we talk about lower-thirds where the input data could be fetched from a web server (like name of speaker etc.).
> 

I was under the impression that we solved that question long time ago and use the schedule.xml by frab/pentabarf. As there is tooling to convert other formats to it it would be a good interchange format. Did i miss a discussion about this?

**Comment by MaZderMind** [comment]
For reference: This seems to be the Code from the Talk (I did not yet watch the Video, just scrolled through the slides): https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/tree/master/ext/wpe

@philn would you mind hopping into the #voctomix irc-channel on hackint to discuss this?
I tried compiling WPEWebKit on Ubuntu 18.04 and had quite some problems. There does not seem to exist a packaged version, if i did not miss anything, and not explicit support for an OS other then Ubuntu 16.04.

**Comment by MaZderMind** [comment]
@derpeter Usually Lower Thirds represent the person(s) currently in the picture. For Multi-Person Talks this would require additional control on top of the information provided by the fahrplan.xml: Which speaker is currently in view and when to show/hide the overlay.

Also we would probably want configurable Styles/Themes per event.
I would rather see this as a separate (Webserver/-Client) application.

**Comment by philn** [comment]
@MaZderMind yeah - WPEWebKit is tricky to build but it should work on a Ubuntu 18.04 host, I believe. It's getting a bit late here, I'll join the IRC channel tomorrow and try to help out :)

**Comment by derpeter** [comment]
> @derpeter Usually Lower Thirds represent the person(s) currently in the picture. For Multi-Person Talks this would require additional control on top of the information provided by the fahrplan.xml: Which speaker is currently in view and when to show/hide the overlay.
> 
> Also we would probably want configurable Styles/Themes per event.
> I would rather see this as a separate (Webserver/-Client) application.

Sure, there was much discussion on how to do this while we signed up for the protoyp found. I would rather not start from 0 again. sytles and themes are the smallest issue as we can just deploy them like the background.png more interesting is if we e.g. want to show the user pictures of frab / pentbarf to make it easy for to operator to identify the person speaking.

**Comment by MaZderMind** [comment]
State of the matter: I was able to extract the Source-Code from @philn's patched gst-plugins-bad repo into a standalone plugin and compile all dependencies under a recent Ubuntu 18.04. See https://github.com/MaZderMind/gst-wpesrc for instructions.

It turns out, that the plugin currently needs a Wayland Execution-Environment and can only be used to render onto the screen, downloading the rendered Frames from the GPU via gldownload is not yet supported.

I was not able to get the Plugin running locally yet.

**Comment by philn** [comment]
We provide a flatpak runtime containing WPEWebKit now, and a gst-wpe demo:
https://github.com/Igalia/gst-wpe-broadcast-demo

**Comment by philn** [comment]
You'll recognize the audio vumeter from voctomix btw, I've ported it to Rust :)

**Comment by philn** [comment]
`gldownload` is supported now btw. You still need a wayland-egl environment though.

**Comment by derpeter** [comment]
I lost a little bit track what we want to achieve here ;-). At least the lower thrid feature is now implemented in voctomix2 as well as support for video4linux devices. This we we could feed the browser window via video4linux loopback e.g. 

ffmpeg ffmpeg -f x11grab -r 15 -s 1920x1080 -i :0.0+0,0 -vcodec rawvideo -pix_fmt yuv420p -threads 0 -f v4l2 /dev/video0.

if i understand correctly this would solve the same issue discussed here. Never the less a gestreamaer wpewebkit source in vocotomixt would be also very nice and should not be much effort with the gsteramer modul in place.

**Comment by philn** [comment]
Is that ffmpeg command doing the html overlay?
A GStreamer-based approach would probably fit more naturally in voctomix, as I understood anyway you already rely on GStreamer for most things?

Anyway, as maintainers it's up to you if you want to use this WPE plugin or not. Feel free to close the issue if you think the use-case is already fulfilled :)

**Comment by derpeter** [comment]
i see no reason not implement also WPE. If you want to try it yourself have a look at the video4linux plugin in the vocotmix2 branch. WPE sould be copy paste and replace the gestreamer element in the pipe.
